src/Microsoft.ML.FastTree/GamClassification.cs

```diff
@@ -50,8 +50,8 @@ public sealed class Arguments : ArgumentsBase
         /// <summary>
         /// Initializes a new instance of <see cref="BinaryClassificationGamTrainer"/>
         /// </summary>
-        internal BinaryClassificationGamTrainer(IHostEnvironment env, Arguments args)
-             : base(env, args, LoadNameValue, TrainerUtils.MakeBoolScalarLabel(args.LabelColumn))
+        internal BinaryClassificationGamTrainer(IHostEnvironment env, Options options)
```

Comment:
One thing that I've noticed about this switch to `options` is that we have two APIs for each learner: One with `options`, and one with a subset of the `options` spelled out. Wouldn't it make sense to only have one constructor that took options, and have the Catalog interface handle the multiplicity of ways to construct it?

That is, only have 
````cs
new MyTrainer(env, options)
````

And then have the catalog do this:

```cs
public static BinaryClassificationGamTrainer GeneralizedAdditiveModels(this BinaryClassificationCatalog.BinaryClassificationTrainers catalog,
            string labelColumn = DefaultColumnNames.Label,
            string featureColumn = DefaultColumnNames.Features,
            string weights = null,
            int numIterations = GamDefaults.NumIterations,
            double learningRate = GamDefaults.LearningRates,
            int maxBins = GamDefaults.MaxBins)
        {
            Contracts.CheckValue(catalog, nameof(catalog));
            var env = CatalogUtils.GetEnvironment(catalog);
            var options = new BinaryClassificationGamTrainer.Options{
                xyz = etc
            }
            return new BinaryClassificationGamTrainer(env, options);
        }
```

Reply:
Indeed.  That's  issue #2100

Recall, we had issues with the *constructor* unification for FastTreeRanking.

P0 is to get the public APIs  ready. 
#2100 is P1/P2  in that context.

---
In reply to: [249970882](https://github.com/dotnet/machinelearning/pull/2199#discussion_r249970882) [](ancestors = 249970882)